src/manager.cpp

```diff
@@ -0,0 +1,225 @@
+#include <xyz/openbmc_project/Session/Item/client.hpp>
+
+#include <session.hpp>
+
+#include <filesystem>
+#include <chrono>
+#include <sstream>
+
+namespace obmc
+{
+namespace session
+{
+
+using namespace obmc::dbus;
+using namespace phosphor::logging;
+using namespace std::chrono_literals;
+
+namespace fs = std::filesystem;
+
+SessionManager::SessionManager(sdbusplus::bus::bus& busIn,
+                               boost::asio::io_context& ioc) :
+    SessionManagerServer(busIn, sessionManagerObjectPath),
+    bus(busIn), ioc(ioc), timer(ioc)
+{
+    dbusManager = std::make_unique<sdbusplus::server::manager::manager>(
+        bus, sessionManagerObjectPath);
+    bus.request_name(serviceName);
+    timer.expires_from_now(10s);
+    timer.async_wait(std::bind(&SessionManager::checkSessionOwnerAlive, this,
+                               std::placeholders::_1));
+}
+
+std::string SessionManager::create(std::string username,
+                                   std::string remoteAddress, SessionType type,
+                                   int32_t callerPid)
+{
+    auto session = this->create(username, remoteAddress, callerPid);
+    if (!session)
+    {
+        throw InvalidArgument();
+    }
+    session->sessionType(type);
+    return session->sessionID();
+}
+
+SessionItemPtr SessionManager::create(const std::string& userName,
+                                      const std::string& remoteAddress,
+                                      pid_t callerPid)
+{
+    auto sessionId = generateSessionId();
+
+    auto sessionObjectPath = getSessionObjectPath(sessionId);
+    auto session =
+        std::make_shared<SessionItem>(bus, sessionObjectPath, callerPid);
+
+    session->sessionID(hexSessionId(sessionId));
+    session->remoteIPAddr(remoteAddress);
+
+    if (!userName.empty())
+    {
+        try
+        {
+            session->adjustSessionOwner(userName);
+        }
+        catch (const std::exception& ex)
+        {
+            log<level::DEBUG>(
+                "Skip publishing the obmcsess object if user not found",
+                entry("USER=%d", userName.c_str()),
+                entry("ERROR=%s", ex.what()));
+            return SessionItemPtr();
+        }
+    }
+
+    sessionItems.emplace(sessionId, session);
+    return session;
+}
+
+uint32_t SessionManager::closeAllByType(SessionType type) 
+{
+    const auto count = std::erase_if(sessionItems, [type](const auto& item) {
+        auto const [id, session] = item;
+        return session->sessionType() == type;
+    });
+
+    return count;
+}
+
+void SessionManager::close(std::string sessionId)
+{
+    log<level::DEBUG>("SessionManager::close()",
+                      entry("SESSION-ID=%d", sessionId.c_str()));
+    try
+    {
+        auto numSessId = parseSessionId(sessionId);
+        if (sessionItems.find(numSessId) == sessionItems.end())
+        {
+            throw InvalidArgument();
+        }
+        sessionItems.erase(numSessId);
```

Comment:
I guess it would be better to store an iterator returned by `find` and use it here to prevent double searching. Or even call `erase(numSessId)` without check. It will return 0 if no items found.

Reply:
Fixed